shell/browser/lib/shutdown_blocker_win.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2020 Microsoft, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+#include <utility>
+
+#include "base/logging.h"
+#include "base/task/post_task.h"
+#include "base/threading/thread_task_runner_handle.h"
+#include "base/win/win_util.h"
+#include "base/win/wrapped_window_proc.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+#include "shell/browser/lib/shutdown_blocker_win.h"
+#include "ui/gfx/win/hwnd_util.h"
+
+namespace electron {
+
+namespace {
+const wchar_t kShutdownBlockerWinWindowClass[] = L"Electron_ShutdownBlockerWin";
+}
+
+ShutdownBlockerWin::ShutdownBlockerWin(bool dedicated_message_loop)
+    : dedicated_message_loop_(dedicated_message_loop),
+      instance_(NULL),
+      window_(NULL) {}
+
+ShutdownBlockerWin::~ShutdownBlockerWin() = default;
+
+void ShutdownBlockerWin::SetShutdownHandler(base::Callback<bool()> handler) {
+  LOG(INFO) << "setting shutdown handler on "
+            << (dedicated_message_loop_ ? "renderer" : "browser") << " process";
+  should_shutdown_ = std::move(handler);
+  BlockShutdown();
+}
+
+void ShutdownBlockerWin::BlockShutdown() {
+  if (blocking_) {
+    return;
+  }
+  blocking_ = true;
+  if (dedicated_message_loop_) {
+    LOG(INFO) << "unblocking shutdown on renderer process";
+    owner_thread_task_runner_ = base::ThreadTaskRunnerHandle::Get();
+    owner_thread_task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(&ShutdownBlockerWin::MessageLoop,
+                                  base::Unretained(this)));
+  } else {
+    LOG(INFO) << "blocking shutdown on browser process";
+    CreateHiddenWindow();
+  }
+}
+
+void ShutdownBlockerWin::MessageLoop() {
+  CreateHiddenWindow();
+  MSG msg;
+  while (GetMessage(&msg, NULL, 0, 0)) {
+    TranslateMessage(&msg);
+    DispatchMessage(&msg);
+  }
+  DestroyHiddenWindow();
+}
+
+void ShutdownBlockerWin::CreateHiddenWindow() {
+  ATOM atom = RegisterWindowClass();
```

Comment:
:atom: ???

Reply:
:electron: instead?

Reply:
The `ATOM` here is a win32 thing: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerclassa.

Reply:
thanks @zcbenz I didn't know that!